channeldb/kvdb/etcd/readwrite_bucket.go

```diff
@@ -0,0 +1,246 @@
+	}
+
+	return newReadWriteBucket(b.tx, makePath(b.path, key))
+}
+
+// CreateBucket creates and returns a new nested bucket with the given
+// key.  Returns ErrBucketExists if the bucket already exists,
+// ErrBucketNameRequired if the key is empty, or ErrIncompatibleValue
+// if the key value is otherwise invalid for the particular database
+// implementation.  Other errors are possible depending on the
+// implementation.
+func (b *readWriteBucket) CreateBucket(key []byte) (kvdb.RwBucket, error) {
+	if len(key) == 0 {
+		return nil, walletdb.ErrBucketNameRequired
+	}
+
+	if !isValidKey(key) {
+		return nil, walletdb.ErrIncompatibleValue
+	}
+
+	bucketKey := makeBucketPrefix(b.path) + string(key)
+	if b.tx.stm.Get(bucketKey) == string(key) {
+		return nil, walletdb.ErrBucketExists
+	}
+
+	b.tx.stm.Put(bucketKey, string(key))
+	return newReadWriteBucket(b.tx, makePath(b.path, key)), nil
+}
+
+// CreateBucketIfNotExists creates and returns a new nested bucket with
+// the given key if it does not already exist.  Returns
+// ErrBucketNameRequired if the key is empty or ErrIncompatibleValue
+// if the key value is otherwise invalid for the particular database
+// backend.  Other errors are possible depending on the implementation.
+func (b *readWriteBucket) CreateBucketIfNotExists(key []byte) (kvdb.RwBucket, error) {
+	bucket, err := b.CreateBucket(key)
+	if err != nil {
+		if err == walletdb.ErrBucketExists {
+			return newReadWriteBucket(b.tx, makePath(b.path, key)), nil
+		}
+
+		return nil, err
+	}
+
+	return bucket, nil
+}
+
+func (b *readWriteBucket) DeleteNestedBucket(key []byte) error {
+
+	// TODO shouldn't empty key return ErrBucketNameRequired ?
+	if len(key) == 0 || !isValidKey(key) {
+		return walletdb.ErrIncompatibleValue
+	}
+
+	bucketKey := makeBucketPrefix(b.path) + string(key)
+	if b.tx.stm.Get(bucketKey) != string(key) {
+		return walletdb.ErrBucketNotFound
+	}
+
+	queue := []string{makePath(b.path, key)}
+
+	for len(queue) != 0 {
+		path := queue[0]
+		queue = queue[1:]
+
+		// Delete values in the current bucket
+		valuePrefix := makeValuePrefix(path)
+		kv := b.tx.stm.First(valuePrefix)
+
+		for kv != nil {
+			b.tx.stm.Del(kv.key)
+			kv = b.tx.stm.Next(valuePrefix, kv.key)
+		}
+
+		// Delete the top bucket and queue sub buckets.
+		bucketPrefix := makeBucketPrefix(path)
+		keySeparator := strings.LastIndex(path, "/")
+		if keySeparator != -1 {
+			// Sub bucket.
+			parent, key := path[:keySeparator], path[keySeparator+1:]
+			bucketToDelete := makeBucketPrefix(parent) + key
+			b.tx.stm.Del(bucketToDelete)
+		} else {
+			// Top level.
+			b.tx.stm.Del(makeBucketPrefix("") + string(key))
+		}
+
+		kv = b.tx.stm.First(bucketPrefix)
+
+		for kv != nil {
+			queue = append(queue, makePath(path, []byte(kv.val)))
+			kv = b.tx.stm.Next(bucketPrefix, kv.key)
+		}
+	}
+
+	return nil
+}
+
+func (b *readWriteBucket) Put(key, value []byte) error {
+	if len(key) == 0 {
+		return walletdb.ErrKeyRequired
+	}
+
+	if !isValidKey(key) {
```

Comment:
since we regularly insert random data, i think we need to have something that converts binary data into a character set that doesn't collide with the prefixing strategy.